eclair-core/src/main/scala/fr/acinq/eclair/io/PeerConnection.scala

```diff
@@ -0,0 +1,486 @@
+            } else if (d.behavior.fundingTxAlreadySpentCount < MAX_FUNDING_TX_ALREADY_SPENT) {
+              d.behavior.copy(fundingTxAlreadySpentCount = d.behavior.fundingTxAlreadySpentCount + 1)
+            } else {
+              log.warning(s"peer sent us too many channel announcements with funding tx already spent (count=${d.behavior.fundingTxAlreadySpentCount + 1}), ignoring network announcements for $IGNORE_NETWORK_ANNOUNCEMENTS_PERIOD")
+              setTimer(ResumeAnnouncements.toString, ResumeAnnouncements, IGNORE_NETWORK_ANNOUNCEMENTS_PERIOD, repeat = false)
+              d.behavior.copy(fundingTxAlreadySpentCount = d.behavior.fundingTxAlreadySpentCount + 1, ignoreNetworkAnnouncement = true)
+            }
+        }
+        stay using d.copy(behavior = behavior1)
+
+      case Event(ResumeAnnouncements, d: ConnectedData) =>
+        log.info(s"resuming processing of network announcements for peer")
+        stay using d.copy(behavior = d.behavior.copy(fundingTxAlreadySpentCount = 0, ignoreNetworkAnnouncement = false))
+    }
+  }
+
+  whenUnhandled {
+    case Event(unhandledMsg: LightningMessage, d: HasTransport) =>
+      // we ack unhandled messages because we don't want to block further reads on the connection
+      d.transport ! TransportHandler.ReadAck(unhandledMsg)
+      log.warning(s"acking unhandled message $unhandledMsg in state $stateName from sender=$sender")
+      stay
+
+    case Event(Terminated(actor), d: HasTransport) if actor == d.transport =>
+      Logs.withMdc(diagLog)(Logs.mdc(category_opt = Some(Logs.LogCategory.CONNECTION))) {
+        log.info("transport died, stopping")
+      }
+      stop(FSM.Normal)
+
+    case Event(_: Rebroadcast, _) => stay // ignored
+
+    case Event(_: DelayedRebroadcast, _) => stay // ignored
+
+    case Event(_: RoutingState, _) => stay // ignored
+
+    case Event(_: TransportHandler.ReadAck, _) => stay // ignored
+
+    case Event(SendPing, _) => stay // we got disconnected in the meantime
+
+    case Event(_: Pong, _) => stay // we got disconnected before receiving the pong
+
+    case Event(_: PingTimeout, _) => stay // we got disconnected after sending a ping
+
+    case Event(_: BadMessage, _) => stay // we got disconnected while syncing
+
+  }
+
+  /**
+   * As long as we receive messages from our peer, we consider it is online and don't send ping requests. If we don't
+   * hear from the peer, we send pings and expect timely answers, otherwise we'll close the connection.
+   *
+   * This is implemented by scheduling a ping request every 30 seconds, and pushing it back every time we receive a
+   * message from the peer.
+   *
+   */
+  def heartbeat(s: StateFunction): StateFunction = {
+    case event if s.isDefinedAt(event) =>
+      event match {
+        case Event(_: LightningMessage, d: ConnectedData) if sender == d.transport =>
+          // this is a message from the peer, he's alive, we can delay the next ping
+          scheduleNextPing()
```

Comment:
`scheduleNextPing()` is only called when we receive a LightningMessage -> if we never received anything once we're in `CONNECTED` state we'll never send a `Ping`

Reply:
Great catch, and missing a test on that. Will add a timeout on the `Init` too.